meshmode/interop/firedrake/mesh.py

```
@@ -0,0 +1,832 @@
+    top.init()
+
+    # Get new element ordering:
+    c_start, c_end = top._topology_dm.getHeightStratum(0)
+    cell_index_mm2fd = np.vectorize(top._cell_numbering.getOffset)(
+        np.arange(c_start, c_end))
+    v_start, v_end = top._topology_dm.getDepthStratum(0)
+
+    # Firedrake goes crazy reordering local vertex numbers,
+    # we've got to work to figure out what changes they made.
+    #
+    # *perm2cells* will map each permutation of local vertex numbers to
+    #              a list of the meshmode cells to which that permutation
+    #              has been applied
+    perm2cells = {}
+    for mm_cell_id, dmp_ids in enumerate(top.cell_closure[cell_index_mm2fd]):
+        # look at order of vertices in firedrake cell
+        vert_dmp_ids = dmp_ids[np.logical_and(v_start <= dmp_ids, dmp_ids < v_end)]
+        fdrake_order = vert_dmp_ids - v_start
+        # get original order
+        mm_order = mesh.groups[group_nr].vertex_indices[mm_cell_id]
+        # want permutation p so that mm_order[p] = fdrake_order
+        # To do so, look at permutations acting by composition.
+        #
+        # mm_order \circ argsort(mm_order) =
+        #     fdrake_order \circ argsort(fdrake_order)
+        # so
+        # mm_order \circ argsort(mm_order) \circ inv(argsort(fdrake_order))
+        #  = fdrake_order
+        #
+        # argsort acts as an inverse, so the desired permutation is:
+        perm = tuple(np.argsort(mm_order)[np.argsort(np.argsort(fdrake_order))])
+        perm2cells.setdefault(perm, [])
+        perm2cells[perm].append(mm_cell_id)
+    perm2cells = {perm: np.array(cells)
+                  for perm, cells in six.iteritems(perm2cells)}
+
+    # Now make a coordinates function
+    from firedrake import VectorFunctionSpace, Function
+    if mesh.is_conforming:
+        family = 'CG'
+    else:
+        warn("Non-conforming meshes are untested,... I think they should work")
```

```suggestion
        warn("Converting non-conforming meshes to Firedrake is untested")
```

---

Better warning to the user?  "Nonconforming" meshes usually mean hanging nodes or other un-aligned bits.  I didn't realize Firedrake used this wording internally.

---

> Better warning to the user? "Nonconforming" meshes usually mean hanging nodes or other un-aligned bits. I didn't realize Firedrake used this wording internally.

`mesh` in the above lines is a [meshmode.mesh.Mesh](https://documen.tician.de/meshmode/mesh.html#meshmode.mesh.Mesh), I don't think Firedrake uses the word internally.

I may have misunderstood what conforming meant, is there a formal definition hanging around somewhere?

---

Looking at the code, perhaps they use the term to mean the coordinates are represented in a CG space rather than a DG one.  I'm sure there is some use case for DG but certainly can't think of it (unless they need it for periodicity)

---

Since I apparently don't quite understand the details of non-conforming meshes, it may be better to just raise an error if the mesh is non-conforming

---

Insisted in https://github.com/inducer/meshmode/pull/29/commits/4c444e363244d43098922df9c34eae5ecfcfb9af